url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+    private void removeEmitter( SessionEmitter sessionEmitter ) throws InterruptedException
+    {
+        SessionEmitterImpl emitterImpl;
+        m_lock.lock();
+        try
+        {
+            emitterImpl = m_emitters.get( sessionEmitter );
+            if (emitterImpl == null)
+                return;
+            /*
+             * m_emitters.remove( sessionEmitter );
+             * It is better to keep the emitter in the container
+             * allowing collider stop properly later.
+             */
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+        emitterImpl.stopAndWait();
+    }
+
+    public void removeEmitterNoWait( final SessionEmitter sessionEmitter )
+    {
+        /* Supposed to be called by SessionEmitterImpl itself. */
+        m_lock.lock();
+        try
+        {
+            m_emitters.remove( sessionEmitter );
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+    }
+
+    public void removeDatagramListenerNoWait( final DatagramListener datagramListener )
+    {
+        /* Supposed to be called by DatagramListenerImpl */
+        m_lock.lock();
+        try
+        {
+            m_datagramListeners.remove( datagramListener );
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+    }
+
+    private static final Logger s_logger = Logger.getLogger( Collider.class.getName() );
+
+    private static final AtomicReferenceFieldUpdater<ColliderImpl, SelectorThreadRunnable> s_strHeadUpdater =
+            AtomicReferenceFieldUpdater.newUpdater( ColliderImpl.class, SelectorThreadRunnable.class, "m_strHead" );
+
+    private static final AtomicReferenceFieldUpdater<ColliderImpl, SelectorThreadRunnable> s_strTailUpdater =
+            AtomicReferenceFieldUpdater.newUpdater( ColliderImpl.class, SelectorThreadRunnable.class, "m_strTail" );
+
+    private static final AtomicReferenceFieldUpdater<SelectorThreadRunnable, SelectorThreadRunnable> s_nextSelectorThreadRunnableUpdater =
+            AtomicReferenceFieldUpdater.newUpdater( SelectorThreadRunnable.class, SelectorThreadRunnable.class, "nextSelectorThreadRunnable" );
+
+    private final Selector m_selector;
+    private final ThreadPool m_threadPool;
+    private boolean m_run;
+
+    private final ReentrantLock m_lock;
+    private final Map<SessionEmitter, SessionEmitterImpl> m_emitters;
+    private final Map<DatagramListener, DatagramListenerImpl> m_datagramListeners;
+    private final Map<Integer, RetainableDataBlockCache> m_dataBlockCache;
+    private RetainableByteBufferPool m_joinPool;
+    private boolean m_stop;
+
+    private volatile SelectorThreadRunnable m_strHead;
+    private volatile SelectorThreadRunnable m_strTail;
+    private SelectorThreadRunnable m_strLater;
+    private final AtomicReference<SelectorAlarm> m_alarm;
+
+    public ColliderImpl( Config config ) throws IOException
+    {
+        super( config );
+
+        m_selector = Selector.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *m\_selector*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).